sdk/python/feast/infra/offline_stores/contrib/mssql_offline_store/mssql.py

```diff
@@ -0,0 +1,641 @@
+            on_demand_feature_views=registry.list_on_demand_feature_views(project),
+        )
+        return job
+
+    @staticmethod
+    def write_logged_features(
+        config: RepoConfig,
+        data: Union[pyarrow.Table, Path],
+        source: LoggingSource,
+        logging_config: LoggingConfig,
+        registry: BaseRegistry,
+    ):
+        raise NotImplementedError()
+
+    @staticmethod
+    def offline_write_batch(
+        config: RepoConfig,
+        feature_view: FeatureView,
+        table: pyarrow.Table,
+        progress: Optional[Callable[[int], Any]],
+    ):
+        raise NotImplementedError()
+
+
+def _assert_expected_columns_in_dataframe(
+    join_keys: Set[str], entity_df_event_timestamp_col: str, entity_df: pandas.DataFrame
+):
+    entity_df_columns = set(entity_df.columns.values)
+    expected_columns = join_keys.copy()
+    expected_columns.add(entity_df_event_timestamp_col)
+
+    missing_keys = expected_columns - entity_df_columns
+
+    if len(missing_keys) != 0:
+        raise errors.FeastEntityDFMissingColumnsError(expected_columns, missing_keys)
+
+
+def _assert_expected_columns_in_sqlserver(
+    join_keys: Set[str], entity_df_event_timestamp_col: str, table_schema: EntitySchema
+):
+    entity_columns = set(table_schema.keys())
+    expected_columns = join_keys.copy()
+    expected_columns.add(entity_df_event_timestamp_col)
+
+    missing_keys = expected_columns - entity_columns
+
+    if len(missing_keys) != 0:
+        raise errors.FeastEntityDFMissingColumnsError(expected_columns, missing_keys)
+
+
+def _get_join_keys(
+    project: str, feature_views: List[FeatureView], registry: BaseRegistry
+) -> Set[str]:
+    join_keys = set()
+    for feature_view in feature_views:
+        entities = feature_view.entities
+        for entity_name in entities:
+            entity = registry.get_entity(entity_name, project)
+            join_keys.add(entity.join_key)
+    return join_keys
+
+
+def _infer_event_timestamp_from_sqlserver_schema(table_schema) -> str:
+    if any(
+        schema_field["COLUMN_NAME"] == DEFAULT_ENTITY_DF_EVENT_TIMESTAMP_COL
+        for schema_field in table_schema
+    ):
+        return DEFAULT_ENTITY_DF_EVENT_TIMESTAMP_COL
+    else:
+        datetime_columns = list(
+            filter(
+                lambda schema_field: schema_field["DATA_TYPE"] == "DATETIMEOFFSET",
+                table_schema,
+            )
+        )
+        if len(datetime_columns) == 1:
+            print(
+                f"Using {datetime_columns[0]['COLUMN_NAME']} as the event timestamp. To specify a column explicitly, please name it {DEFAULT_ENTITY_DF_EVENT_TIMESTAMP_COL}."
+            )
+            return datetime_columns[0].name
+        else:
+            raise ValueError(
+                f"Please provide an entity_df with a column named {DEFAULT_ENTITY_DF_EVENT_TIMESTAMP_COL} representing the time of events."
+            )
+
+
+class MsSqlServerRetrievalJob(RetrievalJob):
```

Comment:
missing to_remote_storage too i guess? so will fail to materialize with materialization engines. might unjust have an explicit call out here

Reply:
fixed